src/server/replica.cc

```diff
@@ -1152,29 +1141,73 @@ error_code Replica::SendCommand(string_view command, ReqSerializer* serializer)
   return ec;
 }
 
-bool Replica::TranactionData::UpdateFromParsedEntry(journal::ParsedEntry&& entry) {
-  if (entry.opcode == journal::Op::EXEC) {
-    shard_cnt = entry.shard_cnt;
-    dbid = entry.dbid;
-    txid = entry.txid;
-    return true;
-  } else if (entry.opcode == journal::Op::COMMAND) {
-    txid = entry.txid;
-    shard_cnt = entry.shard_cnt;
-    dbid = entry.dbid;
-    commands.push_back(std::move(entry.cmd));
-    return true;
-  } else if (entry.opcode == journal::Op::MULTI_COMMAND) {
-    commands.push_back(std::move(entry.cmd));
-    return false;
-  } else {
-    DCHECK(false) << "Unsupported opcode";
+bool Replica::TransactionData::AddEntry(journal::ParsedEntry&& entry) {
+  switch (entry.opcode) {
+    case journal::Op::COMMAND:
+    case journal::Op::EXPIRED:
+      commands.push_back(std::move(entry.cmd));
+      [[fallthrough]];
+    case journal::Op::EXEC:
+      shard_cnt = entry.shard_cnt;
+      dbid = entry.dbid;
+      txid = entry.txid;
+      return true;
+    case journal::Op::MULTI_COMMAND:
+      commands.push_back(std::move(entry.cmd));
+      return false;
+    default:
+      DCHECK(false) << "Unsupported opcode";
   }
   return false;
 }
 
-bool Replica::TranactionData::IsGlobalCmd() const {
+bool Replica::TransactionData::IsGlobalCmd() const {
   return commands.size() == 1 && commands.front().cmd_args.size() == 1;
 }
 
+auto Replica::TransactionReader::Next(JournalReader* reader, Context* cntx)
+    -> optional<TransactionData> {
+  io::Result<journal::ParsedEntry> res;
+  do {
+    if (res = reader->ReadEntry(); !res) {
+      cntx->ReportError(res.error());
+      return std::nullopt;
+    }
+
+    // Expiration is preceding to a MULTI...EXEC sequence, even if it was issued within it.
+    if (res->opcode == journal::Op::EXPIRED) {
+      TransactionData tmp_tx;
+      CHECK(tmp_tx.AddEntry(std::move(*res)));
+      return tmp_tx;
+    }
+
+  } while (!cntx->IsCancelled() && !current_.AddEntry(std::move(*res)));
+
+  return cntx->IsCancelled() ? std::nullopt : make_optional(std::move(current_));
+}
+
+void Replica::TransactionDataQueue::Push(Replica::TransactionData&& tx_data, bool was_inserted,
+                                         const Cancellation* cll) {
+  waker_.await([&]() { return ((queue_.size() < capacity_) || cll->IsCancelled()); });
+  queue_.emplace(std::move(tx_data), was_inserted);
+  waker_.notifyAll();
+}
+
+auto Replica::TransactionDataQueue::Pop(const Cancellation* cll) -> optional<Item> {
+  waker_.await([&]() { return (!queue_.empty() || cll->IsCancelled()); });
+  if (cll->IsCancelled())
+    return std::nullopt;
+
+  DCHECK(!queue_.empty());
+  auto data = std::move(queue_.front());
+  queue_.pop();
+  waker_.notify();  // TODO: maybe notify only after processing?
+
+  return data;
+}
```

Comment:
You execute first, and only them pop from the queue and wake. It seems there is no difference where we pop from the queue, but waking here switches context back to the reader. which might have implications on performance 🤔